File: web/src/lib/encryption/AES.test.ts
import { aesDecrypt, aesEncrypt, NewAESKey } from "./AES";

test("AES", async () => {
	const key = await NewAESKey() as string;

	expect(key.length)
		.toBe(Math.ceil(16/3) * 4);

	const input = "testing";

	let output = await aesEncrypt(key, input) as string;
	expect(output)
		.toHaveLength(Math.ceil((input.length + 12 + 4 + 12)/3) * 4); // IV + counter + append IV


	output = await aesDecrypt(key, output) as string;
	expect(output)
		.toBe(input);

	// Test Go outputs
	const goKey = "9/dd0iJw0TWcg9szIWefsA==";
	const goOutput = "tyOOaSfO6K3bx6lEVikbX7fAnIok3bu1Y9HXNPUZf3kLo3I=";

	output = await aesEncrypt(goKey, input) as string;
	expect(output)
		.toHaveLength(Math.ceil((input.length + 12 + 4 + 12)/3) * 4); // IV + counter + append IV

	output = await aesDecrypt(goKey, output) as string;
	expect(output)
		.toBe(input);

	output = await aesDecrypt(goKey, goOutput) as string;
	expect(output)
		.toBe(input);
});

File: go/crypto/aes_test.go
package crypto

import (
	"math"
	"testing"

	"github.com/candiddev/shared/go/assert"
)

func TestAES(t *testing.T) {
	key, err := NewAESKey()
	assert.Equal(t, err, nil)
	assert.Equal(t, int(math.Ceil(float64(16)/3)*4), len(key))

	input := "testing"

	output, err := EncryptValue(key, input)
	assert.Equal(t, err, nil)
	assert.Equal(t, int(math.Ceil((float64(len(input))+12+4+12)/3)*4), len(output.Ciphertext)) // IV + counter + append IV
	assert.Equal(t, TypeAES128GCM, output.Encryption)

	out, err := output.Decrypt(key)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)

	// Test JavaScript outputs
	key = "lQQyBeRmGYECoYkafl+4VQ=="
	jsOutput, _ := ParseEncryptedValue(string(TypeAES128GCM) + "$Guqhksdu4nsd6FIrONsXnXyDl8wS78amak0uMZ49KJSly5w=")

	output, err = EncryptValue(key, input)
	assert.Equal(t, err, nil)
	assert.Equal(t, int(math.Ceil((float64(len(input))+12+4+12)/3)*4), len(output.Ciphertext)) // IV + counter + append IV

	out, err = output.Decrypt(key)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)

	out, err = jsOutput.Decrypt(key)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)
}
